### ISSUE TITLE
Potential data leakage issue

### ISSUE THREAD

**Issue by minqukanq** [kimanki/TAFAS#4]
I've been studying your TAFAS paper and implementation, specifically focusing on the PAAS (Periodicity-Aware Adaptation Scheduling) approach.
While implementing this method, I noticed what appears to be a data leakage issue in the evaluation procedure.

In tafas.py, the adapt_tafas method seems to access future data points before evaluation:

1. When constructing batches based on the calculated period:
```python
inputs = enc_window_all[batch_start:batch_end], enc_window_stamp_all[batch_start:batch_end], dec_window_all[batch_start:batch_end], dec_window_stamp_all[batch_start:batch_end]
```

2. Then adapting with partial ground truth before evaluation:
```python
pred, ground_truth = self._adapt_with_partial_ground_truth(inputs, period, batch_size, batch_idx)
mse = F.mse_loss(pred, ground_truth, reduction='none').mean(dim=(-2, -1)).detach().cpu().numpy()
```

This sequence appears to:
1. Access t+1 to t+period future data points at time t
2. Adapt the model using this future information
3. Evaluate the adapted model on the same future data
In a true online test-time adaptation scenario.

Could you clarify whether this is intentional or if I'm misunderstanding the implementation? For proper online evaluation, predictions should be made before seeing any future data, and adaptation should occur only with data that would be available at that specific time point.

**Comment by heitorrapela** [comment]
@minqukanq Hello, I am also studying this codebase. Initially, I was thinking the same as you, but if you go inside the function [_adapt_with_partial_ground_truth](https://github.com/kimanki/TAFAS/blob/master/tta/tafas.py#L238-L239), you see that only the partial GT is being used to update the GCM parameters of the model.

**Comment by minqukanq** [comment]
@heitorrapela Thanks for the clarification! One nuance I'd like to highlight is that while only partial ground truth is used, it’s important to note that this "partial" GT still spans the future p-step horizon at the current test time. In effect, the model is updated using future information at the current timestep, which slightly deviates from the conventional notion of strict test-time adaptation.

**Comment by heitorrapela** [comment]
Hi @minqukanq, I agree with you, that this is not exactly the original convention of TTA for other domains, such as computer vision (see my other issue which is related: https://github.com/kimanki/TAFAS/issues/3). But in the paper, their hypothesis is that you can have some lag for annotation parts of the incoming data, which seems to be the case for some papers on data streaming and concept drift. Assumptions to have annotated incoming batches or samples seem okay in time-series adaptation, but of course if you are able to remove any kind of label, and still perform better than the zero-shot you are doing a great job.

**Comment by minqukanq** [comment]
Thank you for the interesting discussion @heitorrapela. I’ve been thinking more carefully about the assumptions made in the paper and how they relate to the current implementation.

If I may say so respectfully, I believe the assumptions made in the paper might differ slightly from how you’ve interpreted them.

The authors did not assume that test data can be used without labels.

The paper does not assume access to future ground truth. Rather, as described in the 4th paragraph of the Introduction ("Interestingly, as depicted..."), it refers to a situation where ground truth values become partially observable after a delay, due to the sequential nature of time series.
In their example, partial ground truth for the first 7 days becomes available only after a full week has passed.
This implies that the model can be adapted using delayed past labels, not future ones.

However, based on the current code implementation, the model appears to use future values — from 𝑡+1 to 𝑡+𝑝  — both to adapt the model and evaluate it on the same data.
This approach involves using information that would not be available at test time, and contradicts the paper’s assumption that only partially observable ground truth becomes available after a delay.

In practice, we cannot access any form of ground-truth whether labeled or not ahead of time.
Therefore, improving zero-shot performance by leveraging any confirmed future values (even if not labeled) is infeasible in a realistic online adaptation scenario.

I'd appreciate your thoughts on this interpretation, especially if I might have missed something from your perspective.

**Comment by heitorrapela** [comment]
Hi, sorry for the late response, I have been a bit busy on my side.

> If I may say so respectfully, I believe the assumptions made in the paper might differ slightly from how you’ve interpreted them.

Response: I complete agree with you, I was saying that if you manage to make it work without labels it would be better for the community (like traditional TTA in computer vision), but the problem is that we cannot relly on entropy minimization like in TENT-based models, because we cannot use normal entropy (like kl-div) in regression, I tried different approaches without labels on my side like ordinary entropy, following the base code of TAFAS, but I did not succed to adapt the GCMs (maybe if you adapt the full-model, but then you lose prior knowledge which can be important depending on what you want to do). You can check in the paper, that they said the same thing (challenges (C)): `C1. Entropy-based TTA losses are infeasible for
regression-based TSF`

> The authors did not assume that test data can be used without labels.

Response: yes, I agree. It is part of their hyp in opportunities (O) `O2. Utilizing partially-observed ground truth enables proactive TSF-TTA.` If you see the following image, the POGT (partial observable GT) is part of the forecast window, so you can see there is a small leak between the one sample per partial partial adaptation and test data for each batch. 

![Image](https://github.com/user-attachments/assets/dbdcf3da-0847-4184-8bdd-1f8cf44f9cd4)

Code for the  [(_adapt_with_partial_ground_truth)](https://github.com/kimanki/TAFAS/blob/master/tta/tafas.py#L238C1-L239C1)
```
            pred_partial, ground_truth_partial = pred[0][:period], ground_truth[0][:period]
            mse_partial = F.mse_loss(pred_partial, ground_truth_partial)
```

They adapt for this one sample: pred[0], and then test on this and the rest of the batch. Even though I agree with your interpretation, I still think that waiting in time to annotate one sample and adapt and predict on the rest can be valuable for real-time applications.

**Comment by minqukanq** [comment]
After carefully reviewing the code, I’ve concluded that there is no data leakage present. 
The authors evaluate the model using the predictions made *before* the adaptation step, which ensures a fair test-time setting.

I appreciate the high-quality work done by the authors and would also like to thank @heitorrapela for taking the time to engage in this thoughtful discussion.

Closing this issue now.
